### PR TITLE
add pillow package

### PR DESCRIPTION
**Pull request by nopid**
A proposal for a simple pillow package.

Support for Pillow 8.0.1 with zlib, jpeg and freetype features thanks to emscripten ports.

A full featured Pillow would require to port lcms, tiff, jpeg2000, imagequant, webp and webpmux.

Closes https://github.com/iodide-project/pyodide/issues/800, Closes https://github.com/iodide-project/pyodide/issues/774

### CODE REVIEW

**Comment by nopid** [comment]
Identified some linking issue with the libraries. Need to be FIXED before going further with PR.

**Comment by nopid** [comment]
Linking issue has been fixed, jpeg files now loading.

**Review by rth — COMMENTED** [review]
Thank you @nopid!

Ideally we should try to make linking work without specifying `extra_objects` manually. For instance, matplotlib uses `USE_FREETYPE=1 -s USE_LIBPNG=1 -s USE_ZLIB=1` without manually appending paths to the libraries when they are used.

File: packages/pillow/patches/setitup.patch [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
++            feature.zlib = "z"
+ 
+         if feature.want("jpeg"):
+             _dbg("Looking for jpeg")
+-            if _find_include_file(self, "jpeglib.h"):
+-                if _find_library_file(self, "jpeg"):
+-                    feature.jpeg = "jpeg"
+-                elif sys.platform == "win32" and _find_library_file(self, "libjpeg"):
+-                    feature.jpeg = "libjpeg"  # alternative name
++            feature.jpeg = "jpeg"
```
[the remark this answers is not in the document]
Comment (nopid) [reply packages/pillow/patches/setitup.patch:39]: Definitely, I will clean it up. My frist PR (8baa632763703a849047029239507dd4874496e4) was using only `-s USE_FREETYPE=1 -s USE_ZLIB=1 -s USE_JPEG=1` as cflags. Unfortunately, it was building without including the static libraries, leading to runtime errors.

The `setup.py` of Pillow is a bit unorthodox and I resolved the issue by using `extra_objects` (the classical way to include static libraries). What code is supposed to translate dynamic library linking into static version invocation?

I noticed that matplotlib also includes one (but only one) of its library by hand using ldflags: 
https://github.com/iodide-project/pyodide/blob/0c91434bb6406605c324566c283838cd29680476/packages/matplotlib/meta.yaml#L25-L27

File: packages/pillow/patches/setitup.patch [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+             defs.append(("HAVE_LIBJPEG", None))
+         if feature.jpeg2000:
+             libs.append(feature.jpeg2000)
+@@ -712,7 +689,8 @@
+             if sys.platform == "win32" and not PLATFORM_MINGW:
+                 defs.append(("OPJ_STATIC", None))
+         if feature.zlib:
+-            libs.append(feature.zlib)
++            #libs.append(feature.zlib)
++            xobj.append("../../../../emsdk/emsdk/.emscripten_cache/asmjs/libz.a")
```
[the remark this answers is not in the document]
Comment (nopid) [reply packages/pillow/patches/setitup.patch:89]: I understand that my handling of dynamic libraries from emscripten ports is suboptimal. Are there some examples of proper use appart from matplotlib?

**Comment by nopid** [comment]
Thank you @rth for the review. I was definitely fixing it wrong. Now the extra_objects are gone and -L and -l take care of static libraries as expected. The only exception is with libjpeg.bc. As a bytecode file and not a library, I understand that the whole  path has to be provided. This was done in ldflags through meta.yaml.

I also added some more test to ensure libjpeg.bc is properly linked.

**Review by rth — COMMENTED** [review]
Thanks @nopid ! Adding libjpeg.bc to ldflags will link to all compiled pillow modules (6 or so) even those that don't need it, which would probably result in some size overhead (unless it's optimized by the linker?). But agree it's simpler and the size is not too bad currently,
```
-rw-r--r-- 1 root root  2.4M Dec 14 20:33 pillow.data
-rw-r--r-- 1 root root   49K Dec 14 20:33 pillow.js
```

LGTM. Will merge unless @dalcde has other comments.

**Comment by dalcde** [comment]
I don't have any comments, but you might want to rebase to current master to make sure it compiles well without `ERROR_ON_MISSING_LIBRARIES=0`.

**Comment by rth** [comment]
@nopid Could you please rebase on master again, to make sure it works with https://github.com/iodide-project/pyodide/pull/843 . Thanks!

**Comment by nopid** [comment]
Suppressed a warning on missing jpeg library and rebased on master.

**Comment by rth** [comment]
@nopid would you be interested in continuing packaging scikit-image https://github.com/iodide-project/pyodide/pull/733 ? having pillow packaged would already help a lot there.

**Comment by nopid** [comment]
> @nopid would you be interested in continuing packaging scikit-image #733 ? having pillow packaged would already help a lot there.

Sure @rth, I will have a look at it and see what I can do.

**Comment by oeway** [comment]
@rth Is this already in the released version of pyodide? Looking forward to try it!

**Comment by rth** [comment]
The master branch is deployed to `https://cdn.jsdelivr.net/pyodide/dev/full/` so you could try it there. Otherwise we'll do a release soon https://github.com/iodide-project/pyodide/issues/831

**Pull request merged**